workspace/src/localization/localization_py/localization_py/EKF.py

```diff
@@ -0,0 +1,200 @@
+import numpy as np
+import math
+import yaml
+
+
+class EKF:
+    """A basic Extended Kalman Filter implementation.
+    
+    This currently uses the 4DOF motion model. The Motion Model inputs are throttle and steering, and the observation model uses GPS and Magnetometer as sensors. States are cartesian coordinates x, y, heading theta, and velocity v.
+
+    Attributes:
+        c_1: The motor resistence torque linear component.
+        c_0: The motor resistence torque constant component.
+        l: The length of the vehicle.
+        r_wheel: The radius of the wheel.
+        i_wheel: the intertia of the wheel.
+        gamma: The gear ratio.
+        tau_0: The stalling torque for the motor torque model.
+        omega_0: The maximum no-load speed for the motor torque model.
+        df_1_dv: The derivative of the scaled motor torque relative to the velocity of the vehicle.
+        Q: The covariane matrix for our states.
+        R: The covariance matrix for our sensors.
+        P: The predicted covariance estimate.
+        dt: The timestep for the filter.
+    """
+
+    def __init__(self, dt):
+        """Initialize the EKF.
+
+        Initialize each of the global variables for the EKF. Load the dynamics parameters for the vehicle from an external YAML file.
+
+        Args:
+            dt: The timestep at which this filter will operate at. 
+        """
+
+        with open('/home/art/art/workspace/src/localization/localization_py/localization_py/4DOF_dynamics.yml', 'r') as yaml_file:
```

Comment:
mmmm hardcoding not a fan of. There are some utilities in ROS to best support that. Alternatively, you could load the yaml as parameters and pass that in. I think that's probs best